# Diferenciação Numérica

A diferenciação numérica lida com o seguinte problema: Dada a função $y = f (x)$ e desejamos obter uma de suas derivadas no ponto $x = x_k$. O termo "dado" significa que temos um algoritmo para computar a função ou possuímos um conjunto de pontos de dados discretos $(x_i, y_i), i = 0, 1,. . . n.$ Em ambos os casos, temos acesso a um número finito de pares (x, y) a partir dos quais calcular a derivada. Se você suspeitar que a diferenciação numérica está relacionada à interpolação, você está correto - um meio de encontrar a derivada é aproximar a função localmente por um polinômio e depois diferenciá-la.

A diferenciação numérica não é um processo particularmente preciso. Ela sofre de um conflito entre erros de arredondamento (causados pela precisão limitada da máquina) e erros inerentes à interpolação. Por esse motivo, uma derivada de uma função nunca pode ser calculada com a mesma precisão da forma analítica.

## Aproximações de Diferenças Finitas

A derivação das aproximações de diferenças finitas para as derivadas de $f (x)$ é baseada em expansões de  [séries de Taylor](https://pt.wikipedia.org/wiki/S%C3%A9rie_de_Taylor) para frente e para trás de f (x) sobre x, tal como:

$$f(x+h) = f(x) + hf'(x) + \frac{h^2}{2!}f''(x) + \frac{h^3}{3!}f'''(x) + \frac{h^4}{4!}f''''(x) + \dots$$

$$f(x-h) = f(x) - hf'(x) + \frac{h^2}{2!}f''(x) - \frac{h^3}{3!}f'''(x) + \frac{h^4}{4!}f''''(x) + \dots$$

Também calculamos as somas e diferenças da série:

$$f(x+h) + f(x-h) = 2f(x) + h^2f''(x) + \frac{h^4}{12}f^{(4)}(x) + \dots $$

$$f(x+h) - f(x-h) = 2hf'(x) + \frac{h^3}{3}f'''(x) + \dots $$

Observe que as somas contêm apenas as derivadas pares, enquanto as diferenças retêm apenas as derivadas ímpares.


## Aproximação por diferenças finitas

### Diferença central


A solução da segunda equação $f(x+h) - f(x-h) = 2hf'(x) + \frac{h^3}{3}f'''(x) + \dots $ para $f'(x)$ é:

$$ f'(x) = \frac{f(x+h) - f(x-h)}{2h} - \frac{h^2}{6}f'''(x) - \dots $$

ou

$$f'(x) = \frac{f(x+h) - f(x-h)}{2h} + \mathcal{O}(h^2)$$

O termo $\mathcal{O}(h^2)$ nos lembra que o erro de truncamento se comporta como $h^2$

Similarmente, a primeira equação:

$$f(x+h) + f(x-h) = 2f(x) + h^2f''(x) + \frac{h^4}{12}f^{(4)}(x) + \dots$$ 

leva a uma aproximação para $f''(x)$

$$f''(x) = \frac{f(x+h) - 2f(x) + f(x-h)}{h^2} + \frac{h^2}{12}f^{(4)}(x) + \dots$$

$$f''(x) = \frac{f(x+h) - 2f(x) + f(x-h)}{h^2} +  \mathcal{O}(h^2)$$

### Diferença para frente e para trás

As aproximações centrais de diferenças finitas nem sempre são utilizáveis. Por exemplo, considere a situação em que a função é dada em $n$ pontos discretos $x_0, x_1, \dots x_n$. Como as diferenças centrais usam valores da função em cada lado de x, não poderíamos calcular as derivadas em $x_0$ e $x_n$. Claramente, há uma necessidade de expressões de diferenças finitas que requerem avaliações da função apenas em um lado de x. Essas expressões são chamadas de aproximações de diferenças finitas para frente e para trás.

Diferenças finitas não-centrais podem ser obtidas de expansões da série de Taylor como :

$$f(x+h) = f(x) + hf'(x) + \frac{h^2}{2!}f''(x) + \frac{h^3}{3!}f'''(x) + \frac{h^4}{4!}f''''(x) + \dots$$

resolvendo para $f'(x)$ obtemos:

$$f'(x) = \frac{f(x+h) - f(x)}{h} - \frac{h}{2}f''(x) - \frac{h^2}{6}f'''(x) \dots  $$

Mantendo apenas o primeiro termo do lado direito temos a primeira aproximação de diferença para frente:

$$f'(x) = \frac{f(x+h) - f(x)}{h} + \mathcal{O}(h)$$

Da mesma forma, a equação:

$$f(x-h) = f(x) - hf'(x) + \frac{h^2}{2!}f''(x) - \frac{h^3}{3!}f'''(x) + \frac{h^4}{4!}f''''(x) + \dots$$

produz a primeira aproximação de diferença para trás:

$$f'(x) = \frac{f(x) - f(x-h)}{h} + \mathcal{O}(h)$$

Observe que o erro de truncamento agora é $\mathcal{O}(h)$, o que não é tão bom quanto  $\mathcal{O}(h^2)$ nas aproximações de diferença centrais.



In [1]:
import numpy as np
import matplotlib.pyplot as plot

## Exemplo

Dada a função $f(x) = 2e^{1.5x}$

1. Calcular $f'(x)$ para $x=3$ utilizando $h = 0.1, 0.05, 0.025$ através usando diferença finita para frente. Calcule o erro real, sabendo que o resultado analítico é $f'(3) = 270.05$


__Definindo a função__ $f(x)$

In [8]:
def fx(x):
    return 2*np.exp(1.5*x)

__Definindo uma função para calcular a diferença finita para frente__

In [9]:
def dff(foo,x,h):
    return (foo(x+h) - foo(x))/h

__Definindo valores do problema__

In [13]:
x = 3
h = np.array([0.1,0.05,0.025])
derivada_f = dff(fx,3,h)
derivada_f

array([291.35708553, 280.43631346, 275.17874869])

__Calculando o erro real__

In [14]:
real = 270.05
erro_r = np.abs(derivada_f - real)/real * 100
erro_r

array([7.89005204, 3.84607053, 1.89918485])

__Criando uma tabela de comparação__

|h|$f'(x)$|$\epsilon_r$ %|
--|-------|------------|
0.1|291.3570|7.890
0.05|280.436|3.846
0.025|275.1787|1.899

Analisando os resultados podemos perceber que o erro cai aproximadamente na mesma proporção da diminuição do $h$, ou seja dividindo o $h$ pela metade o erro diminui aproximadamente pela metade


## Exercícios

1. Repita o exemplo anterior aplicando agora a diferença finita para trás e a diferença finita central, analise o erro real e compare os resultados entre as implementações.

2. Calcule a $f''(x)$ para x = 3.

# Derivadas por interpolação


Se $f(x)$ é dado como um conjunto de pontos discretos, a interpolação pode ser um meio muito eficaz de calcular suas derivadas. A ideia é aproximar a derivada de $f(x)$ pela derivada do interpolante. Este método é particularmente útil se os pontos de dados estiverem localizados em intervalos irregulares de x, quando as aproximações de diferenças finitas listadas não forem aplicáveis.

## Polinômio Interpolante

A ideia aqui é simples: Determine um polinômio de grau $n$:

$$P_{n-1}(x) = a_0 + a_1x+a_2x^2 + \dots + a_nx^n$$

através de n + 1 pontos de dados e, em seguida, avaliar suas derivadas no ponto $x$. Como já discutimos, é geralmente aconselhável limitar o grau do polinômio a menos de quatro, a fim de evitar oscilações espúrias do interpolante. Como essas oscilações são ampliadas com cada diferenciação, seu efeito pode ser devastador.

Em vista desta limitação, a interpolação é geralmente local, envolvendo não mais que alguns pontos de dados de vizinhos mais próximos.

Para pontos de dados uniformemente espaçados, interpolação polinomial e aproximações de diferenças finitas produzem resultados similares. 

## Utilizando regressão polinomial

Muitos dados são obtidos com _ruído_ o que pode levar a erros indesejados. Uma forma de reduzir esse problema é utilizar regressão polinomial para ajustar curvas que sofram menos efeitos de pontos espúrios.

### Exemplo

1. Dados os pontos:

x|$f(x)$
--|--
1.5|1.0628
1.9|1.3961
2.1|1.5432
2.4|1.7349
2.6|1.8423
3.1|2.0397

calcule $f'(2)$ e $f''(2)$ utilizando interpolação polinomial utilizando os três pontos mais próximos.

__Solução__

O polinômio interpolador é $P_2(x) = a_0 + a_1x + a_2x^2$ passando pelos pontos 1.9, 2.1 e 2.4. O sistema de equações para determinar os coeficientes é dado por:

$$\begin{bmatrix}
n & \sum x_i & \sum x_i^2 \\
\sum x_i & \sum x_i^2 & \sum x_i^3 \\
\sum x_i^2 & \sum x_i^3 & \sum x_i^4 
\end{bmatrix}
\begin{bmatrix}
a_0\\
a_1\\
a_2
\end{bmatrix}
=
\begin{bmatrix}
\sum y_i\\
\sum y_i x_i\\
\sum y_i x_i^2
\end{bmatrix}
$$

Depois de substituir os dados, obtemos:
$$\begin{bmatrix}
3 & 6.4 & 13.78 \\
6.4 & 13.78 & 29.944 \\
13.78 & 29.944 & 65.6578 
\end{bmatrix}
\begin{bmatrix}
a_0\\
a_1\\
a_2
\end{bmatrix}
=
\begin{bmatrix}
4.6742\\
10.0571\\
21.8385
\end{bmatrix}
$$

resolvendo o sistema

$$
a=
\begin{bmatrix}
-0.7714\\
1.5075\\
-0.1930
\end{bmatrix}
$$

As derivadas do polinômio são $P'_2(x) = a_1+2a_2x$ e $P''_2(x) = 2a_2$, logo:

$$f'(2) \approx P'_2(x) = 1.5075+2(-0.1930)(2) = 0.7355$$

$$f''(2) \approx P''_2(x) = 2(-0.1930) = -0.3860$$


## Exercício

1. Determine $f'(0)$ e $f'(1)$ a partir dos seguintes dados com ruído:

x|f(x)
---|---
0  | 1.9934
0.2| 2.1465
0.4| 2.2129
0.6| 2.1790
0.8| 2.0683
1.0| 1.9448
1.2| 1.7655
1.4| 1.5891

__Dica 1:__ Utilize um polinômio de ordem 3 para aproximar os dados.

__Dica 2:__ Neste problema, os dados representam $f(x) = (x+2)/cosh(x)$ somado a um ruído aleatório. Logo as derivadas "corretas" são $f'(0) = 1.000$ e $f'(1) = -0.833$